File: server/src/main/scala/producer/message_generator/data_generators/bytes/BytesFromHexGenerator.scala
package producer.message_generator.data_generators.bytes

import com.tools.teal.pulsar.ui.producer.v1.producer as pb
import producer.message_generator.data_generators.string.StringGenerator
import java.util.HexFormat
import org.graalvm.polyglot.Context

case class BytesFromHexGenerator(
    generator: StringGenerator
):
    def generate(polyglotContext: Context): Array[Byte] =
        val hexString = generator.generate(polyglotContext)
        HexFormat.of().parseHex(hexString)

object BytesFromHexGenerator:
    def fromPb(v: pb.BytesFromHexGenerator): BytesFromHexGenerator =
        BytesFromHexGenerator(
            generator = StringGenerator.fromPb(v.generator.get)
        )

    def toPb(v: BytesFromHexGenerator): pb.BytesFromHexGenerator =
        pb.BytesFromHexGenerator(
            generator = Some(StringGenerator.toPb(v.generator))
        )

File: ui/components/ui/ProducerSession/message-generator/data-generators/bytes/bytes-from-hex-generator.ts
import * as pb from '../../../../../../grpc-web/tools/teal/pulsar/ui/producer/v1/producer_pb';
import { StringGenerator, stringGeneratorFromPb, stringGeneratorToPb } from '../string-generator/string-generator';

export type BytesFromHexGenerator = {
  type: 'bytes-from-hex-generator',
  generator: StringGenerator
};

export function bytesFromHexGeneratorFromPb(v: pb.BytesFromHexGenerator): BytesFromHexGenerator {
  return {
    type: 'bytes-from-hex-generator',
    generator: stringGeneratorFromPb(v.getGenerator()!)
  };
}

export function bytesFromHexGeneratorToPb(v: BytesFromHexGenerator): pb.BytesFromHexGenerator {
  const message = new pb.BytesFromHexGenerator();
  message.setGenerator(stringGeneratorToPb(v.generator));

  return message;
}
